core/src/main/scala/org/apache/spark/storage/DiskStore.scala

```diff
@@ -73,17 +81,52 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
   }
 
   def putBytes(blockId: BlockId, bytes: ChunkedByteBuffer): Unit = {
+    val bytesToStore = if (serializerManager.encryptionEnabled) {
+      try {
+        val data = bytes.toByteBuffer
+        val in = new ByteBufferInputStream(data, true)
+        val byteBufOut = new ByteBufferOutputStream(data.remaining())
+        val out = CryptoStreamUtils.createCryptoOutputStream(byteBufOut, conf,
+          serializerManager.encryptionKey.get)
+        try {
+          ByteStreams.copy(in, out)
+        } finally {
+          in.close()
+          out.close()
+        }
+        new ChunkedByteBuffer(byteBufOut.toByteBuffer)
+      } finally {
+        bytes.dispose()
+      }
+    } else {
+      bytes
+    }
+
     put(blockId) { fileOutputStream =>
       val channel = fileOutputStream.getChannel
       Utils.tryWithSafeFinally {
-        bytes.writeFully(channel)
+        bytesToStore.writeFully(channel)
       } {
         channel.close()
       }
     }
   }
 
   def getBytes(blockId: BlockId): ChunkedByteBuffer = {
+    val bytes = readBytes(blockId)
+
+    val in = serializerManager.wrapForEncryption(bytes.toInputStream(dispose = true))
+    new ChunkedByteBuffer(ByteBuffer.wrap(IOUtils.toByteArray(in)))
```

Comment:
Yes, it's easier to have the data encrypted in both places. It also makes performance worse, because there's no reason to encrypt the data in memory. It also requires all sorts of hacks in the block manager, such as the one I added to make the streaming WAL work, which is similar to the the one that would be required to encrypt broadcast data.

Yes, the change to not encrypt in memory and encrypt on disk is larger, more involved, more complex, but it's a better change and results in a block manager that has less caveats (for example, users of the block manager don't need to be aware of encryption at all).

I have ideas on how to fix this, I just haven't had the time to work on them. If you don't feel like trying it out, feel free to close the PR.